### PR TITLE
Promote icu_calendar crate to components and move datetime types into it

### PR DESCRIPTION
**Pull request by Manishearth**
Step 1 of https://github.com/unicode-org/icu4x/issues/1018

We may eventually want a separate crate for these, but this is fine for now.

A lot of stuff is not ideal here, I'm trying to land things bit by bit to avoid bitrotting stuff on large refactors like this.

r? @sffc

### CODE REVIEW

**Comment by zbraniecki** [comment]
Can you elaborate more on "separate crate"? How would a final structure look like between `icu_calendar`, `icu_datetime`, `icu_timezone` and this new crate?

**Review by zbraniecki — APPROVED** [review]
this is mostly code shuffle.

**Comment by Manishearth** [comment]
> Can you elaborate more on "separate crate"? How would a final structure look like between `icu_calendar`, `icu_datetime`, `icu_timezone` and this new crate?

Both would depend on `icu_datetime_types`, basically. I think `icu_calendar` would then depend on `icu_datetime`, implementing its traits

**Comment by Manishearth** [comment]
Anyway, that's not planned in this work (this work is mostly to consolidate the code), that's a potential future fix.

**Comment by zbraniecki** [comment]
I'm confused, if `icu_calendar` is to depend on `icu_datetime` then why the types are to be defined in `icu_calendar` and not in `icu_datetime`, and what would be the value of separating the types into yet another crate?

File: components/calendar/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+/// A list of possible error outcomes for working with various inputs to DateTime inputs
+/// and operations.
+#[derive(Display, Debug)]
+pub enum DateTimeError {
+    /// An input could not be parsed.
+    #[displaydoc("{0}")]
+    Parse(core::num::ParseIntError),
+    /// An input overflowed its range.
+    #[allow(missing_docs)] // TODO(#686) - Add missing docs.
```
Comment (sffc) [inline components/calendar/src/error.rs:18]: Nit: Here and below: why do you need the `#[allow(missing_docs)]` ?

File: components/calendar/src/lib.rs [hunk 2]
```diff
@@ -2,14 +2,19 @@
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
+#![cfg_attr(not(any(test, feature = "std")), no_std)]
+
```
Comment (sffc) [inline components/calendar/src/lib.rs:6]: If we're promoting this crate from experimental to components… can you add module-level docs, or at least make sure there is a ticket blocking 0.4 to track that?  We should keep the components directory "ready to release" as much as possible.

**Review by sffc — APPROVED** [review]

File: components/calendar/src/lib.rs (hunk 2 above)
Comment (gregtatum) [reply components/calendar/src/lib.rs:6]: Suggestion: Similarly, this needs `#![warn(missing_docs)]` enabled for it, with at least suppressions for missing docs and a follow-up bug.

**Review by gregtatum — APPROVED** [review]

File: components/calendar/src/error.rs (hunk 1 above)
Comment (Manishearth) [reply components/calendar/src/error.rs:18]: It's a copy paste, I'll clean that up later

File: components/calendar/src/lib.rs (hunk 2 above)
Comment (Manishearth) [reply components/calendar/src/lib.rs:6]: This is part of my next steps.

Comment (Manishearth) [reply components/calendar/src/lib.rs:6]: (I'm trying to keep this PR to be restricted to code moving)

**Comment by Manishearth** [comment]
> I'm confused, if `icu_calendar` is to depend on `icu_datetime` then why the types are to be defined in `icu_calendar` and not in `icu_datetime`, and what would be the value of separating the types into yet another crate?

I think with _two_ crates it makes sense for datetime to depend on calendar. With three it may make more sense for calendar to depend on datetime and both depend on types. Unsure, again, this is not part of my current plan, it's a future cleanup we can discuss.

Regardless of this _for now_ the dependency tree needs to be this way.

**Pull request merged**